stage1/iottymux/iottymux.go

```diff
@@ -0,0 +1,535 @@
+	flag.StringVar(&appName, "app", "", "Target application name")
+}
+
+// Endpoint represents a single attachable endpoint for an application
+type Endpoint struct {
+	Name     string `json:"name"`
+	Family   string `json:"family"`
+	Protocol string `json:"protocol"`
+	Address  string `json:"address"`
+	Port     string `json:"port"`
+	Flavor   string `json:"flavor"`
+}
+
+// iottymux is a multi-action binary which can be used for:
+//  * creating and muxing a TTY for an application
+//  * proxying streams for an application (stdin/stdout/stderr) over TCP
+//  * listing available attachable endpoints for an application
+func main() {
+	var err error
+	// Parse flag and initialize logging
+	flag.Parse()
+	if os.Getenv("STAGE1_DEBUG") == "true" {
+		debug = true
+	}
+	stage1initcommon.InitDebug(debug)
+	log, diag, _ = rktlog.NewLogSet("iottymux", debug)
+	if !debug {
+		diag.SetOutput(ioutil.Discard)
+	}
+
+	// validate app name
+	_, err = types.NewACName(appName)
+	if err != nil {
+		log.Printf("invalid app name (%s): %v", appName, err)
+		os.Exit(254)
+	}
+
+	var r error
+	switch action {
+	// attaching
+	case "auto-attach":
+		r = actionAttach(true)
+	case "custom-attach":
+		r = actionAttach(false)
+
+	// muxing and proxying
+	case "iomux":
+		statusFile, e := os.OpenFile(filepath.Join(pathPrefix, appName, "endpoints"), os.O_CREATE|os.O_WRONLY|os.O_TRUNC, os.ModePerm)
+		if e != nil {
+			err = e
+			break
+		}
+		r = actionIOMux(statusFile)
+	case "ttymux":
+		statusFile, e := os.OpenFile(filepath.Join(pathPrefix, appName, "endpoints"), os.O_CREATE|os.O_WRONLY|os.O_TRUNC, os.ModePerm)
+		if e != nil {
+			err = e
+			break
+		}
+		r = actionTTYMux(statusFile)
+
+	// listing
+	case "list":
+		fallthrough
+	default:
+		statusFile, e := os.OpenFile(filepath.Join(pathPrefix, appName, "endpoints"), os.O_RDONLY, os.ModePerm)
+		if e != nil {
+			err = e
+			break
+		}
+		r = actionPrint(statusFile, os.Stdout)
+
+	}
+
+	if r != nil {
+		log.PrintE("runtime failure", r)
+		os.Exit(254)
+	}
+	os.Exit(0)
+}
+
+// actionAttach handles the attach action, either in "automatic" or "custom endpoints" mode.
+func actionAttach(autoMode bool) error {
+	// retrieve available endpoints
+	statusFile, err := os.OpenFile(filepath.Join(pathPrefix, appName, "endpoints"), os.O_RDONLY, os.ModePerm)
+	if err != nil {
+		return err
+	}
+	eps, err := getEndpoints(statusFile)
+	_ = statusFile.Close()
+	if err != nil {
+		return err
+	}
+
+	// retrieve custom attach mode
+	var withTTYIn, withTTYOut string
+	var withStdin, withStdout, withStderr string
+	if !autoMode {
+		withTTYIn = os.Getenv("STAGE2_ATTACH_TTYIN")
+		withTTYOut = os.Getenv("STAGE2_ATTACH_TTYOUT")
+		withStdin = os.Getenv("STAGE2_ATTACH_STDIN")
+		withStdout = os.Getenv("STAGE2_ATTACH_STDOUT")
+		withStderr = os.Getenv("STAGE2_ATTACH_STDERR")
+	}
+
+	ec := make(chan error)
+	for s, p := range eps {
+		conn, err := net.Dial("tcp", fmt.Sprintf("127.0.0.1:%s", p))
```

Comment:
Slowly starting to re-consider this choice in favor of a stage1-namespaced file-backed unix socket :blush: